### ISSUE TITLE
[HELP] OIIO uses lots of memory that is never freed

### ISSUE THREAD

**Issue by jochenhz** [OpenImageIO/oiio#3717]
When converting an extensive list of OpenEXR images to PNG, the OIIO lib is using 2 to 3 GB of ram. That itself is not much of an issue, but after the process is done, the memory is still held somewhere in the library.

My code is rather simple:

```
void exr_to_png(...) {
     OIIO::ImageBuf in(inFilename);
     
     OIIO::ROI roi = OIIO::ROI(0, w, 0, h, 0, 1, 0, in.spec().nchannels);
     OIIO::ImageBuf out = OIIO::ImageBufAlgo::resize(in, "", 0, roi);
     out.write(outFilename, OIIO::TypeDesc::UINT8);
}
```

After converting 250+ images with that function (called by 3 worker threads), my process consumes 3 GB of ram that is never freed. What do I have to do to release the allocated memory? I've tried to allocate my own ImageCache, but this is crashing badly in a multithreaded environment. 

Thanks a lot

**Comment by lgritz** [comment]
There are potentially a few things to try.

The first thing I would suspect, as you did, is that the memory is tied up in the ImageCache that quietly backs the ImageBufs -- after you destroy the IB, the pixels may linger in the IC in case they are needed again.

Test 1: after you're all done with the conversion, try:

    ImageCache::create()->invalidate_all(true);

This will get the shared cache (create with no args just gives you the global one) and try to free its resources by telling it that you no longer want to re-use any of the open files or stored pixels.

(Aside: the IC backing the IBs is good when you have the potential of lots of IBs for large images active at the same time, or only need parts of each file you touch, or hop around randomly, or may need the same image multiple times, and you want to do all this with minimal I/O and in a fairly low fixed amount of memory. But in cases where you need each image exactly once, when you need it you will need all the pixels, and you only need a few active at a time and then you're done with each image in turn, then it's probably more efficient to do the "forced" read into an IB-local buffer rather than letting it get backed by a IC.)

Test 2: alternately, if you know you don't want/need your IBs to be backed by an ImageCache, then you can force read into an immediate in-memory representation,

    ImageBuf in(inFilename);
    in.read(0, 0, /*force*/ true);
    ...

that should bypass the IC altogether, I think.

When all is said and done, you may also want to

    std::cout << ImageCache::create()->getstats() << "\n";

and verify that it's not holding any files open or much tile memory.

Now, if THAT doesn't alleviate the problem, then the next thing I would suspect is that somehow the OpenEXR library itself (or PNG?) is retaining some allocations. One way to verify this is to mock up a test where you read all the same files and discard them -- but read them using the ImageInput interface (basically: open, read_image, close) so that you are bypassing ImageBuf and ImageCache entirely. If that, too, appears to be retaining memory, then the culprit is probably in the underlying libOpenEXR.

I guess I should probably have started by asking... precisely how are you determining the memory consumption of your program? Are you aware that most system tools are just reporting the "high water mark" of the most the process ever asked the OS for at any one time, and that subsequent calls to free() don't really return memory to the operating system, but just keep holding it for future malloc calls? (Though on most modern architectures and OS's, it's fairly harmless to have a large virtual address space if you aren't actively touching all the memory. It all goes back to the OS when your process terminates.)

**Comment by lgritz** [comment]
Aside: I, too, find the "does it/doesn't it use an ImageCache underneath" behavior of ImageBuf confusing sometimes. 

It's a trade-off, depending on how you will use the ImageBuf. There is no one right choice, and sometimes I think maybe I got the default wrong. Or, at least, I set up the default to be good for the most demanding cases, instead of the common cases most people will face -- reading just one or couple images, needing the full image of each one, then being finished with it. Maybe that wasn't best.

I have been thinking lately of whether I should change the IB interface so that the caller must explicitly say whether you want the IB to be IC-backed at construction (or reset or read) and kind of force callers to make the choice depending on their anticipated use case, rather than having any default hidden ImageCache side effects.

Thoughts, anyone?

**Comment by jochenhz** [comment]
Hey, thanks a lot for your help.

TL;DR; `invalidate_all` and `in.read(0, 0, true)` does the trick, but:


Based on your suggestions, I reduced the code to just this:

```
void exr_to_png(...) {
     OIIO::ImageBuf in(filepath);
     in.read(0, 0, true);
     OIIO::ImageCache::create()->invalidate_all(true);
     std::cout << OIIO::ImageCache::create()->getstats() << "\n";
     return;
}
```

The last output I am getting is:

```
OpenImageIO ImageCache statistics (shared) ver 2.3.17
Options:  max_memory_MB=1024.0 max_open_files=100 autotile=0
            autoscanline=0 automip=0 forcefloat=0 accept_untiled=1
            accept_unmipped=1 deduplicate=1 unassociatedalpha=0
            failure_retries=0 
  Images : 288 unique
    ImageInputs : 288 created, 0 current, 52 peak
    Total pixel data size of all images referenced : 7.5 GB
    Total actual file size of all images referenced : 1.5 GB
    Pixel data read : 0 B
    File I/O time : 0.2s (0.1s average per thread, for 2 threads)
    File open time only : 0.2s
    Peak cache memory : 0 B
  288 were constant-valued in all pixels
  Top files by bytes read:
  Top files by I/O time:
  Files with slowest I/O rates:
    (nothing took more than 0.25s)
    (fastest was 0.0 MB/s)
  Broken or invalid files: 0
```

So it seems like the ImageCache is still being used. For memory usage tracking I use the Xcode allocation instrument. There it seems the OIIO memory is being freed correctly if I do `invalidate_all`.
When removing the `invalidate_all` call, the memory is not freed, so it seems the ImageCache is indeed still being used internally.
Interestingly, If I remove the call to `in.read(0, 0, true)` we are back to leaking lots of memory (even though I invalidate the shared cache). The leak occurs in `OIIO::ImageBufAlgo::resize` and then in `ImageCacheTile::read()` and it seems like the `invalidate_all` is not freeing up the memory here

**Comment by lgritz** [comment]
Yeah, that's fine -- the ImageCache will still be used to read and cache some header information.

If you eliminate the explicit call to read, it will do so (without knowing to force it to avoid the IC) the first time any other operation needs the pixel values, which will be inside the resize call. So this is all as expected.

You're not "leaking" memory -- it's in the cache, waiting for you to either need the same images again (which it will magically have without needing to re-read the files), or to be evicted from the cache as you need and use new images and exceed the maximum cache size.
